tests/integration/components/bs-modal-test.js

```diff
@@ -246,4 +246,51 @@ module('Integration | Component | bs-modal', function (hooks) {
 
     assert.dom('.modal').isVisible('Modal is visible again');
   });
+
+  testBS3('it animates opening and closing the modal', async function (assert) {
+    this.set('open', false);
+    await render(hbs`<BsModal @open={{this.open}}>Hello world!</BsModal>`);
+
+    this.set('open', true);
+    await waitFor('.modal.show');
+    assert.dom('.modal').hasClass('show');
+
+    await settled();
+    assert.dom('.modal').hasClass('show');
+    assert.dom('.modal').hasClass('in');
+
+    this.set('open', false);
+    await waitUntil(() => {
+      return !find('.modal').classList.contains('in');
+    });
+    assert.dom('.modal').hasClass('show');
+    assert.dom('.modal').doesNotHaveClass('in');
+
+    await settled();
+    assert.dom('.modal').doesNotExist();
+  });
+
+  testBS4('it animates opening and closing the modal', async function (assert) {
+    this.set('open', false);
+    await render(hbs`<BsModal @open={{this.open}}>Hello world!</BsModal>`);
+
+    this.set('open', true);
+    await waitFor('.modal.show');
```

Comment:
The closing part of the test looks good, but not sure about the opening: we basically have to assert that the `.modal` exists first, with `display: 'block'`, but *without* `.show`. And only later `.show` should be added (to trigger CSS transitions, see my other comment). This test would not detect a case where we render the modal with `.show` applied immediately (which would break CSS transitions), right?

I guess this is difficult to test though, as `.show` is added with only a minimal delay (`afterRender`), but `waitFor` is polling the DOM with a much bigger interval, so hard to catch that intermediate state where the modal is in DOM but doesn't have `.show`!? 🤔
(The closing part is easier, as `.show` is removed with a large delay (after the CSS transition))

Reply:
Not sure if I can fully follow. Let me try to rephrase with my own words:

To trigger the transition the step must have two phases:

1. An element exist and is attached to the DOM, which does _not_ have the `.show` class.
2. The `.show class is added to that element.

This is required due to the following code of bootstrap: https://github.com/twbs/bootstrap/blob/6ffb0b48e455430f8a5359ed689ad64c1143fac2/scss/_transitions.scss#L1-L7

I'm not sure if the current implementation takes that into account. I guess it would be best implemented using a modifier, which adds the `.show` class. A modifier is guaranteed to be executed _after_ the element is insert into the DOM. So timing-wise it would be a good fit and implementation shouldn't be too complex.

But testing sounds difficult indeed. We would need to execute an assert _after_ the element is insert into the DOM but _before_ the modifier runs. I'm not sure if that is possible at all.

Reply:
I believe the current implementation (in this PR) is already correct in that regard, isn't it?
* state is set to `opening` (https://github.com/kaliber5/ember-bootstrap/pull/1402/files#diff-e071449aeb5436c622964ec59e95bbb45deb6552acca75046b8f29f479d263bbR465)
* on `afterRender` (when DOM is updated), state transitions to `open` (https://github.com/kaliber5/ember-bootstrap/pull/1402/files#diff-e071449aeb5436c622964ec59e95bbb45deb6552acca75046b8f29f479d263bbR488)

So this is just about testing. And given that you wanted to add tests asserting the correct animation behavior, it seems to me the opening part of it will also pass even with a broken behavior (i.e. inserting into DOM and applying `.show` happens in one step). 

If we can't reasonably test this for the above mentioned reasons, them I am ok with leaving the test as-is, just wanted to point this out.

Reply:
I played a little bit around with it as I'm not an expert with CSS animation. It seems as we not only need to wait until the element is attached to DOM. We also need to wait until browser has time to repaint. The animations have not run if I added the class immediately after attaching element to DOM. Only after delaying it for later with `setTimeout(fn, 0)` CSS animations are playing. Here is a JSFiddle showing it: https://jsfiddle.net/8ycm34df/

I _think_ we get similar timing if scheduling in Ember's `afterRender` runloop queue. But this makes testing even more challenging. Will try to ask in Ember Discuss as it seems to be an interesting one. :smile_cat:

Reply:
Maybe we can use [`transitionend` event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/transitionend_event) to assert that a CSS transition has happened without relying on any implementation details at all. :thinking: Will try to investigate when I have a few minutes left.